### PR TITLE
fix(custom-resources): cannot set logging for state machine generated in CompleteHandler

### LINKED ISSUE #27283: custom resources: Generated step function generates finding due to missing logging
### Describe the bug

When custom resources with provider framework has a completedHandler function it generates a aws step function for polling for completion with a configured interval and total timeout. 

This generated step function however has no logging configuration. Which causes the [StepFunctions.1](https://docs.aws.amazon.com/securityhub/latest/userguide/stepfunctions-controls.html#stepfunctions-1) rule from the [AWS Foundational Security Best Practices (FSBP) standard](https://docs.aws.amazon.com/securityhub/latest/userguide/fsbp-standard.html) set to be triggered. 

This is due to the missing logging configuration. The provider construct however also does not provide any option to configure this logging.

### Expected Behavior

The provider construct adds logging by default following the foundational best practice and/or allows this to be configured.

### Current Behavior

The provider construct does not add logging configuration on step function nor does it allow for it to be configured through construct props. 

### Reproduction Steps

Enable StepFunctions.1 rule.
Create stack with custom resource with both evenHandler and isCompleteHandler to trigger generation of state machine.
Deploy generated template
This template should triggger the rule due to missing logging configuration

### Possible Solution

- Add logging configuration by default (with logRetention from props if provided)
- And/or allow for loggroup (configuration) to be provided to provider construct for use in state machine.

### Additional Information/Context

_No response_

### CDK CLI Version

2.96.0

### Framework Version

_No response_

### Node.js Version

18

### OS

Ubuntu 20.04

### Language

Typescript

### Language Version

_No response_

### Other information

_No response_

### LINKED ISSUE #28577: (custom-resources): Default logging configuration hits Cloudwatch Logs Resource Policy size limit
Please add your +1 👍 to let us know you have encountered this
---

### Overview:

A previous PR enabled the ability to configure logging for a state-machine that was managed via a custom resource. This introduced a bug where the logging resource policy became too large because of the name of the log group. This could break the deploys of existing applications that contained this custom resource.

#### Complete Error Message:

```
@aws-cdk--aws-dynamodb.ReplicaProvider/Provider/waiter-state-machine (Providerwaiterstatemachine5D4A9DF0) Resource handler returned message: "Invalid Logging Configuration: The CloudWatch Logs Resource Policy size was exceeded. We suggest prefixing your CloudWatch log group name with /aws/vendedlogs/states/. (Service: AWSStepFunctions; Status Code: 400; Error Code: InvalidLoggingConfiguration; Request ID: 4ba96f98-4be5-450c-a069-3d4cbf93271a; Proxy: null)"
```

### Workaround:

Lock your cdk version to v1.115.0 or below.

### Solution:

Revert in progress: https://github.com/aws/aws-cdk/pull/28699

## Original Issue:

Title: **(custom-resources): Default logging configuration hits Cloudwatch Logs Resource Policy size limit**
-----------------------
### Describe the bug

Upgrading aws-cdk > 2.115 adds a default logging configuration to custom resources. 

When using `aws_dynamodb.Table` with `replication_regions` specified this automatically generates a `@aws-cdk--aws-dynamodb.ReplicaProvider.NestedStackResource` which contains `Providerwaiterstatemachine` and post upgrade also a new log group

When deploying the following error is returned
`@aws-cdk--aws-dynamodb.ReplicaProvider/Provider/waiter-state-machine (Providerwaiterstatemachine5D4A9DF0) Resource handler returned message: "Invalid Logging Configuration: The CloudWatch Logs Resource Policy size was exceeded. We suggest prefixing your CloudWatch log group name with /aws/vendedlogs/states/. (Service: AWSStepFunctions; Status Code: 400; Error Code: InvalidLoggingConfiguration; Request ID: 4ba96f98-4be5-450c-a069-3d4cbf93271a; Proxy: null)"`

### Expected Behavior

Default logging configuration should be configured in a way to prevent the above error

### Current Behavior

Upgrading to aws-cdk >=2.116 generates a logging configuration with what I believe is no name/prefix specified to the log group

```
"ProviderwaiterstatemachineLogGroupDD693A98": {
   "Type": "AWS::Logs::LogGroup",
   "Properties": {
    "RetentionInDays": 731,
   },
   "UpdateReplacePolicy": "Retain",
   "DeletionPolicy": "Retain",
   "Metadata": {
    "aws:cdk:path": "my-stack/@aws-cdk--aws-dynamodb.ReplicaProvider/Provider/waiter-state-machine/LogGroup/Resource"
   }
  }
```

Given the nested stack is auto generated I see no way to override the properties here.

### Reproduction Steps

Using the below construct with aws-cdk <=2.115 deploys successfully

```
table = aws_dynamodb.Table(
            self,
            "my-table",
            table_name="my-table",
            partition_key={"name": "pk", "type": aws_dynamodb.AttributeType.STRING},
            removal_policy=RemovalPolicy.RETAIN,
            replication_regions=["eu-central-1"],
            billing_mode=aws_dynamodb.BillingMode.PAY_PER_REQUEST,
            encryption=aws_dynamodb.TableEncryption.AWS_MANAGED,
            point_in_time_recovery=True,
        )
```

Upgrading aws-cdk > 2.115 causes the above error due to the newly added log group

### Possible Solution

Applying the prefix recommended in the above error message `/aws/vendedlogs/states/` similar to the recommendation [here] (https://docs.aws.amazon.com/step-functions/latest/dg/cw-logs.html) to the log group name?

### Additional Information/Context

_No response_

### CDK CLI Version

2.116.0

### Framework Version

_No response_

### Node.js Version

18

### OS

Linux

### Language

Python

### Language Version

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by go-to-k**
This PR fixes the case that `StateMachine` generated for `CompleteHandler` in `Provider` cannot set logging.

The default log group name of the state machine automatically created by the `Provider` with `isCompleteHandler` should start with `/aws/vendedlogs/states`. Otherwise, each time the state machine is created, the log resource policy size increases. 

https://docs.aws.amazon.com/step-functions/latest/dg/bp-cwl.html

Closes #27283
Fixes #28577 #28744

Related PR #27310(reverted by #28699), #28587

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/aws-cdk-lib/custom-resources/lib/provider-framework/waiter-state-machine.ts [hunk 1]
```diff
@@ -75,23 +125,73 @@ export class WaiterStateMachine extends Construct {
...
+
+    const logGroup = logOptions?.destination ?? new LogGroup(this, 'LogGroup', {
+      // By using the auto-generated name of the Lambda created in the `Provider` that calls this
+      // `WaiterStateMachine` construct, even if the `Provider` (or its parent) is deleted and then
+      // created again, the log group name will not duplicate previously created one with removal
+      // policy `RETAIN`. This is because that the Lambda will be re-created again with auto-generated name.
+      // The `node.addr` is also used to prevent duplicate names no matter how many times this construct
+      // is created in the stack. It will not duplicate if called on other stacks.
+      logGroupName: `/aws/vendedlogs/states/waiter-state-machine-${this.isCompleteHandler.functionName}-${this.node.addr}`,
+    });
```
Comment (go-to-k) [inline packages/aws-cdk-lib/custom-resources/lib/provider-framework/waiter-state-machine.ts:185]: This code fixes [this issue](https://github.com/aws/aws-cdk/issues/28577).

Comment (go-to-k) [inline packages/aws-cdk-lib/custom-resources/lib/provider-framework/waiter-state-machine.ts:185]: This code fixes [this issue](https://github.com/aws/aws-cdk/issues/28577).

The other codes are basically the same as [this PR](https://github.com/aws/aws-cdk/pull/27310) which has been reverted.

File: packages/aws-cdk-lib/custom-resources/lib/provider-framework/waiter-state-machine.ts [hunk 2]
```diff
@@ -75,23 +125,77 @@ export class WaiterStateMachine extends Construct {
...
+      // https://docs.aws.amazon.com/step-functions/latest/dg/bp-cwl.html
+      //
+      // By using the auto-generated name of the Lambda created in the `Provider` that calls this
+      // `WaiterStateMachine` construct, even if the `Provider` (or its parent) is deleted and then
+      // created again, the log group name will not duplicate previously created one with removal
+      // policy `RETAIN`. This is because that the Lambda will be re-created again with auto-generated name.
+      // The `node.addr` is also used to prevent duplicate names no matter how many times this construct
+      // is created in the stack. It will not duplicate if called on other stacks.
+      logGroupName: `/aws/vendedlogs/states/waiter-state-machine-${this.isCompleteHandler.functionName}-${this.node.addr}`,
+    });
```
Comment (go-to-k) [inline packages/aws-cdk-lib/custom-resources/lib/provider-framework/waiter-state-machine.ts:189]: This code fixes [this issue](https://github.com/aws/aws-cdk/issues/28577).

The other codes are basically the same as [this PR](https://github.com/aws/aws-cdk/pull/27310) which has been reverted.

**Comment by TheRealAmazonKendra** [comment]
I see that we reverted one of the linked PRs. Could you please provide some context on why that one was reverted and how this one differs? I'm always a little more cautious on PRs where we've had a previous revert.

**Comment by go-to-k** [comment]
@paulhcsun 

I fixed the CI failure caused by merging main branch.
This is because snapshots for integ tests that use CompleteHandler in PRs created later than this PR need to be changed.

**Review by paulhcsun — COMMENTED** [review]
The changes look good to me and I've verified locally that all of the integ tests changes deployed successfully. Before I go ahead and approve, is there a new test case that covers the regression from the reverted change or is no longer a concern with the updated change in the `renderLoggingConfiguration` method?

**Comment by go-to-k** [comment]
@paulhcsun 

Thanks for your review.

> Before I go ahead and approve, is there a new test case that covers the regression from the reverted change or is no longer a concern with the updated change in the renderLoggingConfiguration method?

I think [this new test](https://github.com/go-to-k/aws-cdk/blob/d9a67aea5a4e300973b067741fd99358c3425687/packages/aws-cdk-lib/custom-resources/test/provider-framework/waiter-state-machine.test.ts#L266), which checks that the log group name begins with the prefix `/aws/vendedlogs/states`, could cover the regression.

**Review by paulhcsun — DISMISSED** [review]

**Comment by paulhcsun** [comment]
@Mergifyio update

**Comment by paulhcsun** [comment]
@Mergifyio update

**Comment by paulhcsun** [comment]
@mergifyio update

**Review by paulhcsun — APPROVED** [review]

**Comment by GavinZZ** [comment]
@mergify update

**Comment by paulhcsun** [comment]
@mergifyio update

**Pull request merged**